flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java

```diff
@@ -0,0 +1,579 @@
+import org.apache.flink.runtime.query.netty.message.KvStateRequestSerializer;
+import org.apache.flink.runtime.state.KeyGroupRangeAssignment;
+import org.apache.flink.runtime.state.internal.InternalMapState;
+import org.apache.flink.util.Preconditions;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+import org.rocksdb.WriteOptions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * {@link MapState} implementation that stores state in RocksDB.
+ * <p>
+ * <p>{@link RocksDBStateBackend} must ensure that we set the
+ * {@link org.rocksdb.StringAppendOperator} on the column family that we use for our state since
+ * we use the {@code merge()} call.
+ *
+ * @param <K>  The type of the key.
+ * @param <N>  The type of the namespace.
+ * @param <UK> The type of the keys in the map state.
+ * @param <UV> The type of the values in the map state.
+ */
+public class RocksDBMapState<K, N, UK, UV>
+	extends AbstractRocksDBState<K, N, MapState<UK, UV>, MapStateDescriptor<UK, UV>, Map<UK, UV>>
+	implements InternalMapState<N, UK, UV> {
+
+	/** Serializer for the keys and values */
+	private final TypeSerializer<UK> userKeySerializer;
+	private final TypeSerializer<UV> userValueSerializer;
+
+	/**
+	 * We disable writes to the write-ahead-log here. We can't have these in the base class
+	 * because JNI segfaults for some reason if they are.
+	 */
+	private final WriteOptions writeOptions;
+
+	/**
+	 * Creates a new {@code RocksDBMapState}.
+	 *
+	 * @param namespaceSerializer The serializer for the namespace.
+	 * @param stateDesc The state identifier for the state.
+	 */
+	public RocksDBMapState(ColumnFamilyHandle columnFamily,
+			TypeSerializer<N> namespaceSerializer,
+			MapStateDescriptor<UK, UV> stateDesc,
+			RocksDBKeyedStateBackend<K> backend) {
+
+		super(columnFamily, namespaceSerializer, stateDesc, backend);
+
+		this.userKeySerializer = stateDesc.getKeySerializer();
+		this.userValueSerializer = stateDesc.getValueSerializer();
+
+		writeOptions = new WriteOptions();
+		writeOptions.setDisableWAL(true);
+	}
+
+	// ------------------------------------------------------------------------
+	//  MapState Implementation
+	// ------------------------------------------------------------------------
+
+	@Override
+	public UV get(UK userKey) throws IOException {
+		try {
+			byte[] rawKeyBytes = serializeUserKeyWithCurrentKeyAndNamespace(userKey);
+			byte[] rawValueBytes = backend.db.get(columnFamily, rawKeyBytes);
+
+			return (rawValueBytes == null ? null : deserializeUserValue(rawValueBytes));
+		} catch (RocksDBException e) {
+			throw new RuntimeException("Error while getting data from RocksDB.", e);
+		}
+	}
+
+	@Override
+	public void put(UK userKey, UV userValue) throws IOException {
+		if (userValue == null) {
+			remove(userKey);
```

Comment:
I think that this would be a weakness of the serializer, or that some data types simply would not allow for `null` values (`null` would be a programming error for them). My feeling is that this code should not have to work around this problem. Probably any type serializer could trivally support null by returning a byte[0].